File: Solutions/0345- Reverse Vowels of a String/Solution.py
class Solution:
    def reverseVowels(self, s: str) -> str:
        vowels = set("aeiouAEIOU")
        s_list = list(s)
        left, right = 0, len(s) - 1
        
        while left < right:
            if s_list[left] in vowels:
                while right > left:
                    if s_list[right] in vowels:
                        s_list[left], s_list[right] = s_list[right], s_list[left]
                        right -= 1
                        break
                    right -= 1
            left += 1

        return ''.join(s_list)

File: Solutions/0345- Reverse Vowels of a String/Solution.java
class Solution {
    public String reverseVowels(String s) {
        Set<Character> set = new HashSet<>(Arrays.asList('a','e','i','o','u','A','E','I','O','U'));
        int left = 0;
        int right = s.length() - 1;
        StringBuilder newString = new StringBuilder(s);
        while (left < right) {
            if (set.contains(s.charAt(left))) {
                while (right > left) {
                    if (set.contains(s.charAt(right))) {
                        newString.setCharAt(left,s.charAt(right));
                        newString.setCharAt(right,s.charAt(left));
                        right--;
                        break;
                    }
                    right--;
                }
            }
            left++;
        }
        return newString.toString();
    }
}